File: docs/guides/dart/lib/saveObjectsModified.dart
import 'dart:convert';
import 'dart:io';

import 'package:algolia_client_search/algolia_client_search.dart';

void saveObjectsModified() async {
  final client =
      SearchClient(appId: 'ALGOLIA_APPLICATION_ID', apiKey: 'ALGOLIA_API_KEY');

  final path = '/tmp/records.json';
  final json = File(path).readAsStringSync();
  List<Map<String, dynamic>> products;
  try {
    products = List<Map<String, dynamic>>.from(jsonDecode(json));
  } catch (e) {
    throw Exception('Failed to read file at $path: $e');
  }

  final records = products.map((product) {
    final reference = product['product_reference'].toString();
    final suffixes = [
      for (int i = 1; i <= reference.length; i++)
        reference.substring(i - 1, reference.length)
    ];
    final updatedProduct = {
      ...product,
      'product_reference_suffixes': suffixes,
    };
    return updatedProduct;
  }).toList();

  final batchParams = BatchWriteParams(
      requests: records
          .map((record) => BatchRequest(
                action: Action.addObject,
                body: record,
              ))
          .toList());
  await client.batch(
    indexName: "<YOUR_INDEX_NAME>",
    batchWriteParams: batchParams,
  );
}

File: docs/guides/csharp/src/saveObjectsModified.cs
namespace Algolia;

using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text.Json;
using Algolia.Search.Clients;
using Algolia.Search.Http;
using Algolia.Search.Models.Search;

class SaveObjectsModified
{
  async Task Main(string[] args)
  {
    try
    {
      var client = new SearchClient(new SearchConfig("ALGOLIA_APPLICATION_ID", "ALGOLIA_API_KEY"));

      var jsonContent = await File.ReadAllTextAsync("products.json");
      var products = JsonSerializer.Deserialize<List<Dictionary<string, object>>>(jsonContent);

      var records = products
        ?.Select(product =>
        {
          var reference = product.GetValueOrDefault("product_reference", "") as string;
          var suffixes = new List<string>();

          for (var i = reference!.Length; i > 1; i--)
          {
            suffixes.Add(reference[i..]);
          }

          return new Dictionary<string, object>(product) { ["suffixes"] = suffixes };
        })
        .ToList();

      await client.SaveObjectsAsync("<YOUR_INDEX_NAME>", records);
    }
    catch (Exception e)
    {
      Console.WriteLine(e.Message);
    }
  }
}
